[CL-18] toast component and service

willmartian opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [ ] Bug fix
- [x] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Adds a toast component and service to the CL. Remove direct client references to `ngx-toastr`

## Code changes

<!--Explain the changes you've made to each file or major component. This should help the reviewer understand your changes-->
<!--Also refer to any related changes or PRs in other repositories-->


- **apps** 
  - Removed toast css from all clients and consolidated in CL
  - Replaced `ToastrService` with `ToastService` in app components
  - **messages.json:** Added variant name so SR can announce it upon opening 
- **libs/components** 
  - **toast.component:** Presentational component
  - **toastr.component:** `ngx-toastr` wrapper for `bit-toast`
  - **toast.service:** Facade over `ToastrService`
    - removed functionality to use custom HTML in toasts
- **libs/common**
  - deprecated `PlatfromUtilsService.showToast`

## Screenshots

See new stories in Storybook. 

<img width="338" alt="image" src="https://github.com/bitwarden/clients/assets/17113462/33aa39bb-fe74-4ef1-bb64-8cfafc7a4200">

<img width="338" alt="image" src="https://github.com/bitwarden/clients/assets/17113462/fd1e8075-82d5-43c9-b1a1-8d61bacf71e8">


https://github.com/bitwarden/clients/assets/17113462/86f842cb-2a54-4c0c-9768-6b04164bf56d



https://github.com/bitwarden/clients/assets/17113462/583bfb2c-4320-48d6-980b-6ebcb7b2000e



## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

danielleflinn reviewed (dismissed):
Looks good! I couldn't get the Service stories to work on my end, but I assume that is a bug with chromatic not storybook.

Hinton commented on libs/components/src/toast/toastr.component.ts:
```diff
@@ -0,0 +1,27 @@
+import { Toast as BaseToastrComponent } from "ngx-toastr";
+
+@Component({
+  selector: "[toast-component2]",
```
Why a `2`?

willmartian replied:
That what `ngx-toastr` expects. This component was already present, moved from [here.](https://github.com/bitwarden/clients/pull/6490/files#diff-7914c620b5900cf916dca16508529205bf5b551a5f9d1618349018722b04a655L14-L15) (I should have moved it in a separate commit so Git would recognize it as a move.) My main change was using `bit-toast` in the template.

Hinton replied:
The demo of using a pink component doesn't require specific selector https://github.com/scttcper/ngx-toastr#using-a-custom-toast

willmartian replied:
Huh, you are right. Don't know why I thought otherwise. I removed the selector entirely since this only presented by a controller.

willmartian commented on libs/components/src/toast/toast.component.ts:
```diff
@@ -0,0 +1,66 @@
+  /** An optional title to display over the message. */
+  @Input() title: string;
+
+  /**
+   * The percent width of the progress bar, from 0-100
+   **/
+  @Input() progressWidth = 0;
```
Inputs are supplied by the `toastr` [component](https://github.com/bitwarden/clients/pull/6490/files/07c5a515af0e5bccef4255e657cae846fb948840#diff-2a855d1135d1790bd13afd87a9fc12836ea895734fdc120bed2d3fda809329a8R8-R14)

That component extends a base component that the library provides. 

I wanted `bit-toast` to be completely presentation and separate from the library.

willmartian commented on libs/common/src/platform/abstractions/platform-utils.service.ts:
```diff
@@ -28,6 +28,7 @@ export abstract class PlatformUtilsService {
   getApplicationVersionNumber: () => Promise<string>;
   supportsWebAuthn: (win: Window) => boolean;
   supportsDuo: () => boolean;
+  /** @deprecated use `@bitwarden/components/ToastService.showToast` instead */
```
Added epic link: https://github.com/bitwarden/clients/pull/6490/commits/fb381830dc923dca5d498fe4323676b1d7f71391

willmartian commented on libs/components/src/toast/toast.component.html:
```diff
@@ -0,0 +1,22 @@
+    <i aria-hidden="true" class="bwi tw-text-xl tw-py-1.5 tw-px-2.5 {{ iconClass }}"></i>
+    <div>
+      <span class="tw-sr-only">{{ variant | i18n }}</span>
+      <div *ngIf="title" class="tw-font-semibold">{{ title }}</div>
```
I'm unsure what would be better to use here. The [example toasts](https://github.com/scttcper/ngx-toastr/blob/master/src/app/pink.toast.ts#L36-L38) also use `div`. 

I don't _think_ it is valid to use a `h1` etc. 

We could use `p`, but to my knowledge the only a11y advantage of that is SR users can skip from it to the next paragraph, but 1.) that doesn't seem necessary for a short title, and 2.) I don't think paragraphs in aria-live alerts can be skipped.

In short, I don't think choosing something else makes a difference here. @vleague2 what do you think?

vleague2 replied:
I don't think it affects the user's experience, but I guess a `p` is probably semantically more accurate.

willmartian replied:
Update to use `p` in [eba8678](https://github.com/bitwarden/clients/pull/6490/commits/eba8678015516a9855af6ff3e7a2a0c87880b408)

willmartian commented on libs/components/src/toast/toast.component.html:
```diff
@@ -0,0 +1,22 @@
+<div
+  class="tw-mb-1 tw-min-w-[300px] tw-text-contrast tw-flex tw-flex-col tw-justify-between tw-rounded-md tw-pointer-events-auto tw-cursor-default {{
```
We only provide a few variants for min-w, none of which are close to this. 

<img width="265" alt="image" src="https://github.com/bitwarden/clients/assets/17113462/7a801d7b-f9d3-4781-9286-800fcaf9dedf">


We also want to share a single reference to this value between Tailwind and the Toastr css, so I am trying out this token approach: [dfaab89](https://github.com/bitwarden/clients/pull/6490/commits/dfaab8956085f4d4f5f0c3e4c456654d342b0aa6)

willmartian commented on apps/web/src/app/app.component.ts:
```diff
@@ -205,7 +202,7 @@ export class AppComponent implements OnDestroy, OnInit {
           }
           case "showToast":
-            this.showToast(message);
+            this.toastService._showToast(message);
```
I want to do that in a separate PR. (Will create a ticket and add to this comment)

I think most calls to `PlatformUtilsService.showToast` already occur in a foreground context.

willmartian commented:
> @danielleflinn  Looks good! I couldn't get the Service stories to work on my end, but I assume that is a bug with chromatic not storybook.

Yeah, I think it might be some Chromatic bug. Works fine locally.

vleague2 commented on libs/components/src/toast/toast.module.ts:
```diff
@@ -0,0 +1,39 @@
+  }
+}
+
+export const BitwardenToastrGlobalConfig: GlobalConfig = {
```
How would a consumer modify things like auto dismiss for individual toasts? Would we pass it in via the service?

willmartian replied:
Any options that we _want_ to be modifiable at the individual toast level could be added to the `ToastOptions` type and passed in [here.](https://github.com/bitwarden/clients/pull/6490/files/07c5a515af0e5bccef4255e657cae846fb948840#diff-3d71c0041b7d0c25959ecd5ecb4457c587fcb399d7b2fea4375e7260946b2fa3R30-R32) So we could update the service method to allow for the following:

```ts
toastService.showToast({ ... autoDismiss: false });
```

However, one goal of a design system is to limit the amount of control that consumers have in favor of opinionated defaults. _Should_ a consumer be able to alter the auto dismiss behavior, or should it be the same for all toasts?

danielleflinn replied:
@willmartian If the toast service can calculate a long enough time limit for a user to read the full toast message, than I would be ok with not allowing the consumer alter the dismiss behavior. Right now the rule design uses is to define auto-dismiss behavior for accessibility, is: `a minimum of 5 seconds + 1 extra second per 120 additional words.`

willmartian replied:
@danielleflinn `autoDismiss` controls whether the toast closes when the timer finishes--I think we always want this to be the case. I believe you are referring to `timeout`, which controls how long the toast timer is. 

> `a minimum of 5 seconds + 1 extra second per 120 additional words`

I can update the service to calculate this and set it as the default timeout. Are there instances where we might not want to follow this rule?

danielleflinn replied:
I don't think there is any instance where we want the default action to not be close. 

Updating the default timer to calculate this would be great! 

I think there might be instances that may require an override, one that came up recently for me was for the browser passkey flow and adding an excluded domain. It felt too long to keep the toast up for a full 5 seconds before closing the window. I think we opted for 2 or 3 seconds in that instance.

willmartian replied:
Sounds good! Added the auto timer in [94b337e](https://github.com/bitwarden/clients/pull/6490/commits/94b337e6228797913c6907d6ac02e45379bf8e97)

vleague2 replied:
👍  I am all for consistent behavior 😎

vleague2 commented on libs/angular/src/components/toastr.component.ts:
```diff
@@ -1,98 +0,0 @@
-      <div
-        *ngIf="message && options.enableHtml"
-        role="alertdialog"
-        aria-live="polite"
```
So `ngx-toast` already had `aria-live` on it, and we were applying it again here?

willmartian replied:
Yes, that is my understand of this https://github.com/scttcper/ngx-toastr/issues/933

vleague2 reviewed (commented):
Some small questions!

vleague2 reviewed (dismissed):
It looks like all feedback has been addressed except for adding more comments to `toastr.css` I believe, but otherwise looks good.

justindbaur reviewed (dismissed):
Looks great, just had a few thoughts about the calculation of the toast timeout but also have no idea how it could be made more locale aware because as far as I know there is no good standardization on "reading speed per character" for locales.

willmartian commented on libs/components/src/toast/utils.ts:
```diff
@@ -0,0 +1,14 @@
+ * @param message the toast message to be displayed
+ * @returns the timeout length in milliseconds
+ */
+export const calculateToastTimeout = (message: string | string[]): number => {
+  const paragraphs = Array.isArray(message) ? message : [message];
+  const numWords = paragraphs
+    .map((paragraph) => paragraph.split(/\s+/).filter((word) => word !== ""))
+    .flat().length;
+  return 5000 + Math.floor(numWords / 120) * 1000;
+};
```
@danielleflinn what do you think?

danielleflinn replied:
I agree with Justin, the limit is likely a bit English-centric but we did pull it form WCAG criteria. I'm not sure how to go about solving it though. Maybe based on the clients selected language we could make the character count dynamic. 

All that said though, toast messages should be pretty short from a design perspective, so cases where we need another 120 characters should be fairly low. Furthermore, Will doesn't this PR include the behavior where if a user mouses over the toast the timer will stop? 

Maybe we leave this as is for now and if we see user feedback about users not having the time to read the full toast, we can iterate on the character count.

willmartian replied:
> @danielleflinn doesn't this PR include the behavior where if a user mouses over the toast the timer will stop?

Yes, with that in mind, I agree we can wait and see if someone reports the default time as a problem. It seems like something that would be fixable on a case by case level rather than prescriptively here and now. 

---

As for including action/links in toasts: I lean towards toasts being more of an escape hatch for errors. If an error is complex, and benefits from additional user action, we should probably display with something other than a toast in the first place. 

That being said, I am happy to add something like this in another PR if we deem it necessary, but I would want to support it without utilizing raw HTML.

justindbaur replied:
Awesome, if this number comes from WCAG that is more than satisfactory. I would also hope that very few of our toasts go longer than 120 characters of any locale. Shipping this and keep an ear on the ground works great for me.

@willmartian Very good point, if there is an error we know enough about to write docs about it, then there is likely a much better way to handle that error in app and provide a better experience in that case.

willmartian replied:
> than 120 characters

Is the rule 120 characters or words? @danielleflinn

justindbaur replied:
Oh sorry, my brain skipped over the `/\s+/` in the function, you're doing words not characters and the previous discussion was about words. I was the one that brought up characters, the fact that it's words likely mean this works even better in other locales. Sorry.

danielleflinn replied:
"Words" is what we have specced out in Figma sorry for not catching the mix up earlier

Hinton commented on libs/components/src/toast/toastr.component.ts:
```diff
@@ -0,0 +1,26 @@
+      (onClose)="remove()"
+    ></bit-toast>
+  `,
+  animations: [
+    trigger("flyInOut", [
+      state("inactive", style({ opacity: 0 })),
+      state("active", style({ opacity: 1 })),
+      state("removed", style({ opacity: 0 })),
+      transition("inactive => active", animate("{{ easeTime }}ms {{ easing }}")),
+      transition("active => removed", animate("{{ easeTime }}ms {{ easing }}")),
+    ]),
+  ],
```
Would it make more sense for the animation to be in the presentational component? If the goal was to decouple it from the toastr library?

willmartian replied:
The animation is [applied](https://github.com/scttcper/ngx-toastr/blob/master/src/lib/toastr/toast.component.ts#L64-L67) by the base `Toast` component that this extends from. As such, it made sense to me to keep it part of this wrapper.

willmartian commented:
I wasn't properly passing the `title` from the service to the component, fixed in [2e30155](https://github.com/bitwarden/clients/pull/6490/commits/2e30155d79fd187ebe786c72d9497117e044ab16).

willmartian commented:
Latest commits fix the conflicts with https://github.com/bitwarden/clients/pull/8211

willmartian commented:
- Added test ids at QA's request: [ef6be25](https://github.com/bitwarden/clients/pull/6490/commits/ef6be2521262b223399dc8a1cbe0f33a7028f419)
- Fixed a bug on desktop where hovering the toast didn't pause it: [b83d668](https://github.com/bitwarden/clients/pull/6490/commits/b83d668faa390557b81fc1b0fd63eff429bd631f)
- Updated colors to align with recent palette changes: [2a52e6b](https://github.com/bitwarden/clients/pull/6490/commits/2a52e6bb21e329f12f55e8eda998cc1ab2b284b6)

vleague2 commented on libs/components/src/toast/toast.component.html:
```diff
@@ -7,8 +7,10 @@
-      <p *ngIf="title" class="tw-font-semibold tw-mb-0">{{ title }}</p>
-      <p *ngFor="let m of messageArray" class="tw-mb-2 last:tw-mb-0">{{ m }}</p>
+      <p *ngIf="title" data-testid="toast-title" class="tw-font-semibold tw-mb-0">{{ title }}</p>
+      <p *ngFor="let m of messageArray" data-testid="toast-message" class="tw-mb-2 last:tw-mb-0">
```
Do we need unique test IDs for each message in the array?

willmartian replied:
No, to my understanding QA just needed a query selector to distinguish between title and message. They can query `[data-testid=toast-message]` to get a query list of all of the messages.
